deepspeed/runtime/zero/stage_1_and_2.py

```diff
@@ -730,36 +733,38 @@ def independent_gradient_partition_epilogue(self):
             # It is safe to clear previously reduced grads of other partitions
             self._clear_previous_reduced_grads()
 
+        self.accumulate_grads(after_reduction=True)
+        see_memory_usage(f"End ipg_epilogue")
+
+    def accumulate_grads(self, after_reduction=False):
         if self.cpu_offload is False:
             for i, _ in enumerate(self.bit16_groups):
-
                 if not i in self.averaged_gradients or self.averaged_gradients[i] is None:
                     self.averaged_gradients[i] = self.get_flat_partition(
                         self.params_in_partition[i],
                         self.first_offset[i],
                         self.partition_size[i],
-                        dtype=self.dtype,
+                        dtype=self.grad_accum_dtype,
                         device=get_accelerator().current_device_name(),
                         return_tensor_list=True)
                 else:
                     avg_new = self.get_flat_partition(
                         self.params_in_partition[i],
                         self.first_offset[i],
                         self.partition_size[i],
-                        dtype=self.dtype,
+                        dtype=self.grad_accum_dtype,
                         device=get_accelerator().current_device_name(),
                         return_tensor_list=True)
 
                     for accumulated_grad, new_avg_grad in zip(self.averaged_gradients[i], avg_new):
                         accumulated_grad.add_(new_avg_grad)
 
-        self._release_ipg_buffers()
-
-        # No need to keep the gradients anymore.
-        # All gradients required by the step
-        # are in self.averaged_gradients
-        self.zero_grad(set_to_none=True)
-        see_memory_usage(f"End ipg_epilogue")
+        if after_reduction:
+            self._release_ipg_buffers()
+            # No need to keep the gradients anymore.
+            # All gradients required by the step
+            # are in self.averaged_gradients
+            self.zero_grad(set_to_none=True)
```

Comment:
@tjruwase In Stage 1 should we wait to zero param.grad until after the reduction?

Reply:
Yes